File: binary search.cpp
// Bismillahir Rahmanir Raheem
#include <bits/stdc++.h>
using namespace std;

int binary_search(int a[],int n, int x)
{
    int mid,left,right;
    left=0;
    right=n-1;
    while(left <=right)
    {
        mid=(left+right)/2;
        if(a[mid]==x)
            return mid;
        if(a[mid]<x)
            left = mid+1;
        else
            right = mid -1;
    }

    return -1;
}

int main()
{
    int a[1000],elements,x;
    scanf("%d",&elements);
    for(int i=0;i<elements;i++)
        scanf("%d",&a[i]);


    scanf("%d",&x);

   int ans = binary_search(a,elements,x);
     if(ans==-1)
        printf("not found");
     else
        printf("%d",ans);

     return 0;

}

File: Search/binary_search.py
from typing import List


class BS:
    def __init__(self,  nums: List[int]):
        nums.sort()
        self.nums = nums

    def binary_search(self, target: int) ->int:

        return self.recursive_binary_search(self.nums, target, 0, len(self.nums) - 1)

    def recursive_binary_search(self, nums: List[int], target:int, left:int, right: int) -> int:

        if left > right:
            return -1

        mid = (left + right) // 2

        if target < nums[mid]:
            return self.recursive_binary_search(nums, target, left, mid - 1)
        if target > nums[mid]:
            return self.recursive_binary_search(nums, target, mid + 1, right)

        return mid


bs = BS([0,-1,3,9,5,12])
print(bs.binary_search(9))

bs = BS([0,-1,3,9,5,12])
print(bs.binary_search(2))
